OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs

```diff
@@ -629,6 +630,165 @@ public static DreamValue NativeProc_clamp(DreamObject instance, DreamObject usr,
             }
             return new DreamValue(0);
         }
+        
+        [DreamProc("gradient")]
+        [DreamProcParameter("A", Type = DreamValueType.DreamObject)]
+        [DreamProcParameter("index", Type = DreamValueType.Float)]
+        public static DreamValue NativeProc_gradient(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
+            if (arguments.ArgumentCount <= 1) {
+                throw new IndexOutOfRangeException("Not enough arguments");
+            }
+
+            /// We dont want keyword arguments screwing with this
+            DreamValue indx = arguments.GetArgument(arguments.OrderedArguments.Count - 1, "index");
+
+            int colorspace = 0;
+            bool loop = false;
+
+            if (!indx.TryGetValueAsFloat(out float index)) {
+                throw new FormatException("Failed to parse index as float");
+            }
+
+            List<DreamValue> GradientList;
+
+            if (arguments.GetArgument(0, "A").TryGetValueAsDreamList(out DreamList gradlist)) {
+                GradientList = gradlist.GetValues();
+
+                if (gradlist.ContainsValue(new DreamValue("loop"))) { loop = true; }
+
+                DreamValue dictspace = gradlist.GetValue(new("space"));
+                if (dictspace != DreamValue.Null && dictspace.TryGetValueAsInteger(out int dictlookup)) { colorspace = dictlookup; }
+
+            } else {
+                GradientList = arguments.GetAllArguments();
+            }
+
+            if (arguments.OrderedArguments.Contains(new("loop"))) { loop = true; }
+
+            if (arguments.NamedArguments.TryGetValue("space", out DreamValue namedlookup) && namedlookup.TryGetValueAsInteger(out int space)) { colorspace = space; }
+
+            /// true: look for int: false look for color
+            bool color_or_int = true;
+
+            float workingfloat = 0;
+            float maxvalue = 1;
+            float minvalue = 0;
+            float leftboundy = 0;
+            float rightboundry = 1;
+
+            /// None of these should be null however C# is "special"
+            Color left = new();
+            Color right = new();
+
+            bool leftexists = false;
+            bool rightexists = false;
+
+            foreach (DreamValue value in GradientList) {
+                if (color_or_int && value.TryGetValueAsFloat(out float flt)) { // Int
+                    color_or_int = false;
+                    workingfloat = flt;
+                    maxvalue = Math.Max(maxvalue, flt);
+                    minvalue = Math.Min(minvalue, flt);
+                    continue; // Succesful parse
+                }
+
+                if (value.TryGetValueAsString(out string? strvalue) && ColorHelpers.TryParseColor(strvalue, out Color color)) {
```

Comment:
`strvalue` should be renamed to `strValue`.

This should also treat anything that isn't a valid color (even something like `new /obj()`) as black instead of ignoring it. That is what BYOND does.

`gradient(0, "white", 0.5, new /obj(), 1, "white", 0.75)` is `#7f7f7f7f`, halfway between black and white.